Added 'STYLE' file containing a proposed style guide for Arbalest

LorenzoPegorari opened the pull request:
Proposal of a style guide for Arbalest's code. This proposed style guide is basically a simplified version of the current BRL-CAD style guide described in the "HACKING" file.

Differently from BRL-CAD's style guide, this one:
- defines only the extensions .cpp (C++ files) and .h (header files)
- doesn't define the code organization (for now)
- doesn't define the build system (for now)
- defines for language compliance C++17 as the standard language ([C++17 is the C++ minimum version for Qt6](https://www.qt.io/blog/qt6-development-hosts-and-targets))
- doesn't define `NEAR_EQUAL` and `NEAR_ZERO` (they aren't necessary)
- doesn't define the _libbu_ functions (they aren't necessary)
- doesn't require emacs and vi local variables block setting at the end of each file
- doesn't define a naming convention for public API functions (not necessary)
- allow "//" style comments

LorenzoPegorari commented:
Also it's important to note that after a style guide is accepted, I personally will clean up the code to follow these new rules.

drossberg commented:
@LorenzoPegorari, I value your effort, but I won't approve it. I don't like the rules, especially the mixed tab/white space one. However, I don't make the decisions anyway.

LorenzoPegorari commented:
Thank you so much to both of you for your honest feedback!

Initially, accordingly to the idea of making arbalest the main BRL-CAD's GUI and making MOOSE the main BRL-CAD's interface, I thought that one of the goals was to also make the style guide for these 3 projects as uniform as possible.

Tbh, I also wouldn't really enjoy these rules, but I thought that having a cohesive style rule was gonna be more important... I really need to learn to be less afraid of proposing something new :)

Having said that, more then anything, my objective was to get some honest feedback. I knew that this proposal wasn't gonna be accepted immediately. Maybe I should have started a discussion somewhere else, but I really thought that showing an actual "real proposal through a PR was the best choice.

LorenzoPegorari commented:
Now, getting to the point, I updated the PR, opting for the `clang-format` idea.

I spent a lot of time (much more than what I would like to admit) trying to understand which clang-format version to use, but I really didn't get anywhere. In the end I opted for version `18`.

`clang-format` has 7 standard styles. I tested all of them to check which one was the closest one to the current _arbalest_ code. The testing was done using `MainWindow.cpp` (the biggest file), changed it's style to each of these standard styles, and checked for differences using `git diff`. These are the results:
- standard style _GNU_ = 2327 different lines
- standard style _LLVM_ = 2207 different lines
- standard style _Chromium_ = 2201 different lines
- standard style _Google_ = 2154 different lines
- standard style _WebKit_ = 995 different lines
- standard style _**Microsoft**_ = 906 different lines

The Microsoft standard style is obviously the closest one to the current code, and it is the one that I used.

I also modified the STYLE file accordingly to these new `clang-format` rules, and I also committed the modified version of `MainWindow.cpp`, so that everyone can see what the new style is gonna look like.

brlcad commented:
@LorenzoPegorari belated reply, but this does look better overall. Biggest deviation from the existing style appears to be pointer alignment.  How much change is there if you  customize DerivePointerAlignment (true) and PointerAlignment (Left)?  The rest of the changes I saw in the diff appeared to mostly be useful inconsistency cleanup.

drossberg commented:
Hmm, all upper-case symbols are for macros, or?

drossberg commented:
Generally, I like to talk about coding rules, maybe not here but on Zulip because of the greater audience. I'm very busy at the moment, but I plan to post my current set of rules on Zulip as start for a discussion.
